Moddable terrain improvements

esbudylin opened the pull request:
This PR continues the work on Terrain Improvements from #710. Previously, all terrain improvements were hard-coded in the game code. This PR moves terrain improvement definitions into the JSON file, which makes them moddable. 

This PR also moves the tile yield bonuses  into the TerrainImprovement class: previously, they were part of the TerrainType class which allowed defining bonuses only for mine, irrigation, and road improvements. With the new system, a bonus yield of any type can be defined for any terrain improvement. 

To fully match the Civ3 bonuses, a special tileModifier field was introduced to the TerrainImprovement class. Similar to the building class (#697), it holds a delegate loaded from a Lua function. Its goal is to allow bonus yields depending on tile context: for the Civ3 ruleset, this is useful for implementing the railroad bonus (only grant additional yield if a mine or irrigation is present on a tile).

Finally (the size of this PR got out of hand!), I made a few changes to the Terraform class. Now terraforms (worker's jobs) can be associated directly with a terrain improvement. Also, the existing logic for terraform effects and prerequisites was moved to Lua.

Right now we have almost all the elements for moddable terrain improvements: it is possible to define new terrain improvements, worker jobs for them, and associate them with custom graphics. The only missing system is the UI for worker jobs: right now the game displays buttons only for a hard-coded list of worker actions. Making it moddable would be a logical step forward.

esbudylin committed:
Update TileOverlayLayer: display any Terrain Improvements

Add zIndex field to the TerrainImprovement class

esbudylin committed:
Move Terraform rules to Lua

esbudylin committed:
Terrain Improvements data moved to the save file

Data on improvement yield bonuses moved from the TerrainType to
TerrainImprovement

esbudylin committed:
Tile Yield modifier for railroads

esbudylin committed:
Merge branch 'Development' into yegor/terrain-improvements-lua

TomWerner commented on C7Engine/C7GameData/TerrainImprovement.cs:
```diff
@@ -8,20 +11,14 @@ public enum Layer {
 		public readonly StrengthBonus? defenseBonus;
 
+		// On the game map, overlapping terrain improvement with a higher zIndex cover those with a smaller one
+		public readonly int zIndex = 0;
```
How does this differ from the layer?

esbudylin replied:
zIndex only affects the drawing order on the game map. The layer affects the gameplay (each tile can have only one type of improvement per layer). Two improvements can belong to the same layer, but have different zIndexes (e.g. irrigation is drawn before roads, and mine is drawn after roads). I added a comment with an explanation to the layer field.

TomWerner commented on C7Engine/C7GameData/TerrainImprovement.cs:
```diff
@@ -32,24 +29,58 @@ public enum Layer {
+		private Dictionary<(TerrainType, Tile.YieldType), int> bonusYields = [];
+		private readonly SaveTerrainImprovement dataSource;
+
 		public TerrainImprovement(
```
Do we need both constructors?

TomWerner replied:
Edit: is one for the hardcoded civ3 rule list?

esbudylin replied:
The first constructor is used in EdgeWalkerTest to create a test road instance

TomWerner commented on C7/Lua/rules/civ3/terraforms.lua:
```diff
@@ -0,0 +1,39 @@
+
+local terraforms = {
+  build_mine = {
+    validator = function(context)
```
I found it a little confusing at first that we have some tables here with a validator, and some with an effect, but none with both.

I think that's just a naming artifact, right? Would naming these like irrigate_validator/mine_validator/etc work, or would that break things?

If it would break things, it might be worth the duplication to duplicate the clear_foliage validator so the wetlands/forests have a validator and an effect, and then have a comment in mine/irrigate that the default effect is to add the improvement.

esbudylin replied:
I agree, this table should be simpler. There's no need for each function to have a sub-table. Fixed it.

TomWerner commented on C7/Lua/rules/civ3/terrain_improvements.lua:
```diff
@@ -0,0 +1,26 @@
+local terrain_improvements = {
+  railroad = {
+    tile_modifier = function(yield)
+      local production_improvement = yield.tile.overlays:ImprovementAtLayer(Layer.ResourceDevelopment)
```
maybe "yield_improvement"? I don't normally think of irrigation as a production improvement

esbudylin replied:
Good catch, done.

TomWerner commented on C7/Lua/rules/civ3.lua:
```diff
@@ -1,67 +1,5 @@
-  end,
+return {
+  buildings = require "civ3.buildings",
+  terraforms = require "civ3.terraforms",
```
It might be worth a comment in the Lua file explaining what the difference between a terraform and a terrain improvement is.

esbudylin replied:
I added a comment for now, but I think terraform is just a bad name for this type of entity. We should name it something clearer like "worker job".

TomWerner commented on C7Engine/C7GameData/Save/SaveTerrainImprovement.cs:
```diff
@@ -0,0 +1,72 @@
+namespace C7GameData.Save;
+
+public class SaveTerrainImprovement {
+	public static IEnumerable<SaveTerrainImprovement> Civ3Improvements() {
```
Could you leave a comment explaining why we need this list of improvements?

esbudylin replied:
Done

TomWerner reviewed (approved):
This is a pretty big PR so I probably missed something, but this is super cool!

I stuck this in the discord, but this allows for crazy worker moves like "clear hill"

<img width="545" height="559" alt="image" src="https://github.com/user-attachments/assets/b8ff9360-352f-4eba-9456-26441f993da5" />

<img width="331" height="178" alt="image" src="https://github.com/user-attachments/assets/9104bcb7-cf4f-47f9-ac51-43e2811b9c1a" />

<img width="331" height="178" alt="image" src="https://github.com/user-attachments/assets/26f9163d-e1a0-41f5-83dd-be9c97d8941b" />

esbudylin committed:
Address review comments

esbudylin committed:
Simplify Lua terraforms table

esbudylin commented:
@TomWerner Thanks for the review! I think I've addressed all your comments. I'll leave this PR open for a few days, in case anyone else wants to contribute a review.
